File: layout21tetris/src/bbox.rs
//!
//! # Rectangular Bounding Boxes
//!

// Crates.io
use serde::{Deserialize, Serialize};

// Local imports
use crate::coords::{HasUnits, Xy};
use crate::placement::Side;

/// # Bounding Rectangular Box
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct BoundBox<T: HasUnits> {
    pub p0: Xy<T>,
    pub p1: Xy<T>,
}
impl<T: HasUnits> BoundBox<T> {
    /// Create a new [BoundBox] from two [Xy] points.
    pub fn new(p0: Xy<T>, p1: Xy<T>) -> Self {
        Self { p0, p1 }
    }
    /// Retrieve our coordinate at [Side] `side`.
    pub fn side(&self, side: Side) -> T {
        match side {
            Side::Left => self.p0.x,
            Side::Right => self.p1.x,
            Side::Bottom => self.p0.y,
            Side::Top => self.p1.y,
        }
    }
}
impl<T: HasUnits + std::cmp::PartialOrd> BoundBox<T> {
    /// Create a new [BoundBox] from potentially unordered pairs of x and y coordinates.
    pub fn from_xy(xs: (T, T), ys: (T, T)) -> Self {
        let (x0, x1) = if xs.0 < xs.1 {
            (xs.0, xs.1)
        } else {
            (xs.1, xs.0)
        };
        let (y0, y1) = if ys.0 < ys.1 {
            (ys.0, ys.1)
        } else {
            (ys.1, ys.0)
        };
        Self::new(Xy::new(x0, y0), Xy::new(x1, y1))
    }
}

/// Trait for types that can be converted to a [BoundBox].
///
/// Includes a single method `boundbox` which returns a [BoundBox] of the type.
pub trait HasBoundBox {
    type Units: HasUnits;
    type Error;
    /// Get a [BoundBox] of the type.
    fn boundbox(&self) -> Result<BoundBox<Self::Units>, Self::Error>;
}

File: Tetris/tetris/bbox.py
#
# # Rectangular Bounding Boxes
#

from typing import TypeVar, Generic, Tuple

from pydantic.generics import GenericModel

# Local imports
from .coords import Xy
from .side import Side


T = TypeVar("T")


class BoundBox(GenericModel, Generic[T]):
    """# Rectangular Bounding Box"""

    mins: Xy[T]  # Minimum (x,y) coordinates
    maxs: Xy[T]  # Maximum (x,y) coordinates

    @property
    def top(self) -> T:
        return self.maxs.y

    @property
    def bot(self) -> T:
        return self.mins.y

    @property
    def bottom(self) -> T:
        return self.mins.y

    @property
    def left(self) -> T:
        return self.mins.x

    @property
    def right(self) -> T:
        return self.maxs.x

    def side(self, side: Side) -> T:
        """# Retrieve our coordinate at [Side] `side`."""
        if side == Side.Left:
            return self.mins.x
        if side == Side.Right:
            return self.maxs.x
        if side == Side.Bottom:
            return self.mins.y
        if side == Side.Top:
            return self.maxs.y
        raise ValueError(f"Invalid side: {side}")

    def from_xy(xs: Tuple[T, T], ys: Tuple[T, T]) -> "BoundBox":
        """# Create a new [BoundBox] from potentially unordered pairs of x and y coordinates."""
        (x0, x1) = (xs[0], xs[1]) if xs[0] < xs[1] else (xs[1], xs[0])
        (y0, y1) = (ys[0], ys[1]) if ys[0] < ys[1] else (ys[1], ys[0])
        return BoundBox(Xy(x0, y0), Xy(x1, y1))
